numpy/core/src/multiarray/flagsobject.c

```diff
@@ -195,78 +196,87 @@ arrayflags_dealloc(PyArrayFlagsObject *self)
-            NPY_ARRAY_C_CONTIGUOUS, carray)
+#define CHECK_FOR_FLAG_OR_FLAG(FLAG1, FLAG2) \
+    (flags & (FLAG1)) == (FLAG1) || (flags & (FLAG2)) == (FLAG2)
 
-static PyObject *
-arrayflags_forc_get(PyArrayFlagsObject *self)
-{
-    PyObject *item;
+#define CHECK_FOR_FLAG_AND_NOT_FLAG(FLAG1, FLAG2) \
+    (flags & (FLAG1)) == (FLAG1) && (flags & (FLAG2)) != (FLAG2)
 
-    if (((self->flags & NPY_ARRAY_F_CONTIGUOUS) == NPY_ARRAY_F_CONTIGUOUS) ||
-        ((self->flags & NPY_ARRAY_C_CONTIGUOUS) == NPY_ARRAY_C_CONTIGUOUS)) {
-        item = Py_True;
-    }
-    else {
-        item = Py_False;
+#define GETTER_FROM_FLAG(name, COND) \
+    static PyObject* \
+    arrayflags_##name##_get(PyArrayFlagsObject *self) \
+    { \
+        int flags = _get_flags(self); \
+        if (COND) { \
+            Py_RETURN_TRUE; \
+        } \
+        Py_RETURN_FALSE; \
     }
-    Py_INCREF(item);
-    return item;
-}
 
-static PyObject *
-arrayflags_fnc_get(PyArrayFlagsObject *self)
-{
-    PyObject *item;
-
-    if (((self->flags & NPY_ARRAY_F_CONTIGUOUS) == NPY_ARRAY_F_CONTIGUOUS) &&
-        !((self->flags & NPY_ARRAY_C_CONTIGUOUS) == NPY_ARRAY_C_CONTIGUOUS)) {
-        item = Py_True;
-    }
-    else {
-        item = Py_False;
-    }
-    Py_INCREF(item);
-    return item;
-}
+#define GETTER_FROM_FUNC_OR_FLAG(name, CFUNC, COND) \
+    static PyObject* \
+    arrayflags_##name##_get(PyArrayFlagsObject *self) \
+    { \
+        PyObject *arr = self->arr; \
+        if (arr != NULL) { \
+            if (CFUNC((PyArrayObject *)arr)) { \
+                Py_RETURN_TRUE; \
+            } \
+        } \
+        else { \
+            int flags = self->flags; \
+            if (COND) { \
+                Py_RETURN_TRUE; \
+            } \
+        } \
+        Py_RETURN_FALSE; \
+    }
+
+GETTER_FROM_FUNC_OR_FLAG(c_contiguous, PyArray_IS_C_CONTIGUOUS,
+                         CHECK_FOR_FLAG(NPY_ARRAY_C_CONTIGUOUS));
+GETTER_FROM_FUNC_OR_FLAG(f_contiguous, PyArray_IS_F_CONTIGUOUS,
+                         CHECK_FOR_FLAG(NPY_ARRAY_F_CONTIGUOUS));
+GETTER_FROM_FLAG(updateifcopy, CHECK_FOR_FLAG(NPY_ARRAY_UPDATEIFCOPY));
+GETTER_FROM_FLAG(owndata, CHECK_FOR_FLAG(NPY_ARRAY_OWNDATA));
+GETTER_FROM_FUNC_OR_FLAG(aligned, PyArray_ISALIGNED,
+                         CHECK_FOR_FLAG(NPY_ARRAY_ALIGNED));
+GETTER_FROM_FUNC_OR_FLAG(writeable, PyArray_ISWRITEABLE,
+                         CHECK_FOR_FLAG(NPY_ARRAY_WRITEABLE));
+GETTER_FROM_FUNC_OR_FLAG(behaved, PyArray_ISBEHAVED,
+                         CHECK_FOR_FLAG(NPY_ARRAY_BEHAVED));
+GETTER_FROM_FUNC_OR_FLAG(carray, PyArray_ISCARRAY,
+                         CHECK_FOR_FLAG(NPY_ARRAY_CARRAY));
+GETTER_FROM_FUNC_OR_FLAG(farray, PyArray_ISFARRAY,
+                         CHECK_FOR_FLAG(NPY_ARRAY_FARRAY));
+GETTER_FROM_FUNC_OR_FLAG(forc, PyArray_ISONESEGMENT,
+                         CHECK_FOR_FLAG_OR_FLAG(NPY_ARRAY_F_CONTIGUOUS,
+                                                NPY_ARRAY_C_CONTIGUOUS));
```

Comment:
Isn't this much more readable as `CHECK_FOR_FLAG(NPY_ARRAY_F_CONTIGUOUS) || CHECK_FOR_FLAG(NPY_ARRAY_C_CONTIGUOUS)`?